Lib/pathlib/_abc.py

```diff
@@ -564,14 +565,33 @@ def samefile(self, other_path):
         return (st.st_ino == other_st.st_ino and
                 st.st_dev == other_st.st_dev)
 
-    def _samefile_safe(self, other_path):
+    def _ensure_different_file(self, other_path):
         """
-        Like samefile(), but returns False rather than raising OSError.
+        Raise OSError(EINVAL) if both paths refer to the same file.
         """
         try:
-            return self.samefile(other_path)
+            if not self.samefile(other_path):
+                return
         except (OSError, ValueError):
-            return False
+            return
+        err = OSError(errno.EINVAL, "Source and target are the same file")
+        err.filename = str(self)
+        err.filename2 = str(other_path)
+        raise err
+
+    def _ensure_distinct_path(self, other_path):
+        """
+        Raise OSError(EINVAL) if the other path is within this path.
+        """
+        if self == other_path:
+            err = OSError(errno.EINVAL, "Source and target are the same path")
+        elif self in other_path.parents:
+            err = OSError(errno.EINVAL, "Source path is a parent of target path")
+        else:
+            return
+        err.filename = str(self)
+        err.filename2 = str(other_path)
```

Comment:
(I think you meant `os.fspath()` rather than `os.fsdecode()`? The latter accepts bytes.)

`PurePathBase` and `PathBase` are not `os.PathLike`; if they were, then `open(TarPath('README.md', archive=tarfile.open(...)))` would attempt to open a file called `README.md` on the local filesystem, despite the `TarPath` representing files within an archive, and sadness would follow. This applies to most functions that accept `os.PathLike` arguments - they expect the return value to be a _local_ filesystem path, not some abstract thing.

I vacillated on whether to support non-`str` types in `PurePathBase` as the "raw" path, but I eventually decided against it. And so currently `str(path)` is the correct way to get a string path from an instance of `PurePathBase`.

Reply:
Ah! I'm sorry! I thought it was in the ParserBase class but it was not in this class. I knew about the PurePathBase that assume its paths to be strings but I thought it was in the ParserBase class (due to GH not showing me the context correctly...), hence my question (and yes, I meant fspath, not fsdecode, sorry). Anyway, the `str` is indeed the correct call!